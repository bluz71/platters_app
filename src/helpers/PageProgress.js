import NProgress from 'nprogress'

class PageProgress {
  constructor () {
    this.loaded = false
  }

  start () {
    if (!this.loaded) {
      NProgress.start()
    }
  }

  done () {
    if (!this.loaded) {
      this.loaded = true
      NProgress.done()
    }

    return true
  }
}

export default PageProgress
