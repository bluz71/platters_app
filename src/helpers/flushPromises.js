// Details:
//   https://blog.rescale.com/testing-promise-side-effects-with-asyncawait
//   https://github.com/facebook/jest/issues/2157#issuecomment-279171856

const flushPromises = () => new Promise(resolve => setImmediate(resolve));

export default flushPromises;
