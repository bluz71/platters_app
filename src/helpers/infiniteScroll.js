// Code snippet based on:
//   http://blog.sodhanalibrary.com/2016/08/detect-when-user-scrolls-to-bottom-of.html
//
const infiniteScoll = (pagination, onPageEnd) => {
  const moreToScroll = pagination && pagination.next_page;
  if (!moreToScroll) {
    return;
  }

  const windowHeight = window.innerHeight;
  const body = document.body;
  const html = document.documentElement;
  const docHeight =
    Math.max(
      body.scrollHeight,
      body.offsetHeight,
      html.clientHeight,
      html.scrollHeight,
      html.offsetHeight
    ) - 20;
  const windowBottom = windowHeight + window.pageYOffset;

  if (windowBottom >= docHeight) {
    onPageEnd();
  }
};

export default infiniteScoll;
