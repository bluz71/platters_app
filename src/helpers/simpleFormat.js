// Code snippet provided by:
//   https://makandracards.com/makandra/1395-simple_format-helper-for-javascript
//
const simpleFormat = str => {
  str = str.replace(/\r\n?/, '\n')
  str = str.trim()

  if (str.length > 0) {
    str = str.replace(/\n\n+/g, '</p><p>')
    str = str.replace(/\n/g, '<br />')
    str = `<p>${str}</p>`
  }

  return str
}

export default simpleFormat
