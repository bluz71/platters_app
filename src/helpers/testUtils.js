import { appAuth } from '../lib/appAuth';

// Details:
//   https://blog.rescale.com/testing-promise-side-effects-with-asyncawait
//   https://github.com/facebook/jest/issues/2157#issuecomment-279171856
const flushPromises = () => new Promise(resolve => setImmediate(resolve));

const logInUser = () => (
  appAuth.logIn('eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyIjo1MywibmFtZSI6ImRlbm5pc2I1NSI\
sImFkbWluIjpmYWxzZSwiZXhwIjoxNTM3ODQ4NDA0LCJpc3MiOiJwbGF0dGVycyIsImF1ZCI6InBs\
YXR0ZXJzX2FwcCJ9.ZUV69wWVXfiiMtQ2ZFPvgqvhZWwQ4WC1bI15Q9DpFCU')
);

const logOutUser = () => (
  appAuth.logOut()
);

export { flushPromises, logInUser, logOutUser };
