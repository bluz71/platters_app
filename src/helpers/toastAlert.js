import { toast } from 'react-toastify';

const toastAlert = (message) => {
  toast.error(message, { className: 'ToastAlert' });
};

export default toastAlert;
