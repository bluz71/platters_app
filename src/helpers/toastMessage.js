import { toast } from 'react-toastify';

export const toastAlert = (message) => {
  toast.error(message, { className: 'ToastAlert' });
};

export const toastNotice = (message) => {
  toast.success(message, { className: 'ToastNotice' });
};
