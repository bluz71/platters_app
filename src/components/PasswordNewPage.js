import React, { Component } from 'react'
import axios from 'axios'
import {
  Row,
  Col,
  PageHeader,
  Well,
  Form,
  FormGroup,
  ControlLabel,
  FormControl,
  Button
} from 'react-bootstrap'
import FontAwesome from 'react-fontawesome'
import { API_HOST, APPLICATION_HOST } from '../config'
import { toastAlert, toastNotice } from '../helpers/toastMessage'

const PASSWORD_CHANGE_ENDPOINT = `${API_HOST}/api/passwords`

class PasswordNew extends Component {
  constructor (props) {
    super(props)

    document.title = 'Platters App'

    this.state = {
      submitButtonText: 'Submit'
    }

    // Bind 'this' for callback functions.
    this.handleResetSubmit = this.handleResetSubmit.bind(this)
  }

  handleResetSubmit (event) {
    event.preventDefault()

    this.setState({
      submitButtonText: (
        <div>
          <FontAwesome name='spinner' spin pulse /> Submitting...
        </div>
      )
    })

    const passwordReset = {
      password_reset: {
        email_address: this.emailAddressInput.value,
        application_host: APPLICATION_HOST
      }
    }

    this.postPasswordReset(passwordReset)
  }

  postPasswordReset (passwordReset) {
    axios
      .post(PASSWORD_CHANGE_ENDPOINT, passwordReset)
      .then(response => {
        this.setState({ submitButtonText: 'Submit' })
        toastNotice(
          'You will receive an email within the next few minutes. It contains instructions for changing your password.'
        )
        this.props.history.push('/')
      })
      .catch(error => {
        this.setState({ submitButtonText: 'Submit' })
        if (error.response && error.response.status === 404) {
          toastAlert(
            'Invalid email address, no user with that email address is registered'
          )
        } else {
          toastAlert('Server error, please try again later')
        }
      })
  }

  renderExplanation () {
    return (
      <p>
        To be emailed a link to reset your password, please enter your email
        address.
      </p>
    )
  }

  renderForm () {
    return (
      <Well>
        <Form horizontal onSubmit={this.handleResetSubmit}>
          <ul className='list-group' />

          <FormGroup>
            <Col componentClass={ControlLabel} md={2}>
              Email address
            </Col>
            <Col md={9}>
              <FormControl
                type='email'
                className='email'
                inputRef={input => (this.emailAddressInput = input)}
              />
            </Col>
          </FormGroup>

          <FormGroup>
            <Col mdOffset={2} md={9}>
              <Button
                type='submit'
                bsStyle='success'
                bsSize='small'
                className='submit'
              >
                {this.state.submitButtonText}
              </Button>
            </Col>
          </FormGroup>
        </Form>
      </Well>
    )
  }

  render () {
    return (
      <Row className='PasswordNew'>
        <Col md={10} mdOffset={1}>
          <PageHeader>Reset your password</PageHeader>
          {this.renderExplanation()}
          {this.renderForm()}
        </Col>
      </Row>
    )
  }
}

export default PasswordNew
