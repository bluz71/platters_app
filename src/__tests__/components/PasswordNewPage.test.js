import React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { flushPromises } from '../../helpers/testUtils';
import * as toastModule from '../../helpers/toastMessage';
import PasswordNewPage from '../../components/PasswordNewPage';

describe('<PasswordNewPage />', () => {
  it('it renders without crashing', () => {
    shallow(<PasswordNewPage />);
  });

  it('with valid email address', async () => {
    const spyAlert = jest.spyOn(toastModule, 'toastAlert');
    const spyNotice = jest.spyOn(toastModule, 'toastNotice');
    const wrapper = mount(
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <PasswordNewPage />
      </MemoryRouter>
    );
    wrapper.find('input.email').instance().value = 'fred@example.com';
    wrapper.find('button').simulate('submit');
    await flushPromises();
    expect(spyAlert).not.toHaveBeenCalled();
    expect(spyNotice).toHaveBeenCalled();
    spyAlert.mockRestore();
    spyNotice.mockRestore();
  });

  it('with invalid email address', async () => {
    const spyAlert = jest.spyOn(toastModule, 'toastAlert');
    const spyNotice = jest.spyOn(toastModule, 'toastNotice');
    const wrapper = mount(
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <PasswordNewPage />
      </MemoryRouter>
    );
    wrapper.find('input.email').instance().value = 'peter@example.com';
    wrapper.find('button').simulate('submit');
    await flushPromises();
    expect(spyAlert).toHaveBeenCalled();
    expect(spyNotice).not.toHaveBeenCalled();
    spyAlert.mockRestore();
    spyNotice.mockRestore();
  });
});
