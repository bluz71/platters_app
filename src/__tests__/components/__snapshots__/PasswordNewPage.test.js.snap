// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PasswordNewPage /> renders password reset form 1`] = `
<Row
  bsClass="row"
  className="PasswordNew"
  componentClass="div"
>
  <Col
    bsClass="col"
    componentClass="div"
    md={10}
    mdOffset={1}
  >
    <PageHeader
      bsClass="page-header"
    >
      Reset your password
    </PageHeader>
    <p>
      To be emailed a link to reset your password, please enter your email address.
    </p>
    <Well
      bsClass="well"
    >
      <Form
        bsClass="form"
        componentClass="form"
        horizontal={true}
        inline={false}
        onSubmit={[Function]}
      >
        <ul
          className="list-group"
        />
        <FormGroup
          bsClass="form-group"
        >
          <Col
            bsClass="col"
            componentClass={[Function]}
            md={2}
          >
            Email address
          </Col>
          <Col
            bsClass="col"
            componentClass="div"
            md={9}
          >
            <FormControl
              bsClass="form-control"
              className="email"
              componentClass="input"
              inputRef={[Function]}
              type="email"
            />
          </Col>
        </FormGroup>
        <FormGroup
          bsClass="form-group"
        >
          <Col
            bsClass="col"
            componentClass="div"
            md={9}
            mdOffset={2}
          >
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsSize="small"
              bsStyle="success"
              className="submit"
              disabled={false}
              type="submit"
            >
              Submit
            </Button>
          </Col>
        </FormGroup>
      </Form>
    </Well>
  </Col>
</Row>
`;
